File: make_heatmap.m
%creates a false color heatmap NAI values predicted by a regression at each pixel in an image, for each image in a folder
%trained regression must be loaded into workspace first
%heatmap images are saved as .png files to folder containing original images
%to distinguish heatmap images, "Heatmap_" is added to beginning of the original file name

myFolder = '\\client\c$\Users\bca08_000\Desktop\Testing\Test'; %replace with filepath to folder where images are saved
if ~isdir(myFolder)
  errorMessage = sprintf('Error: The following folder does not exist:\n%s', myFolder);
  uiwait(warndlg(errorMessage));
  return;
end
filePattern = fullfile(myFolder, '*.png');
images = dir(filePattern);

color_space = 'Name of color space'; %replace with name of color space which regression was trained with data from
%e.g. sRGB, HSV, YIQ, L*a*b*, or YCbCr

%create array to store average predicted NAI for each image
avg_NAI = cell(length(images) + 1, 2);
avg_NAI{1, 1} = "Image Name";
avg_NAI{1, 2} = "Average Predicted NAI";

%analyzes all files with extension .png in myFolder
%images are read in numeric order based on name -> starts at lowest and counts up
for image_count = 1:1:length(images)
    baseFileName = images(image_count).name;
    fprintf('Now reading %s\n', baseFileName)
    fullFileName = fullfile(myFolder, baseFileName);
    RGB_image = imread(fullFileName);
    [height, width, pages] = size(RGB_image);
    leaf_pixel_count = 0;
    leaf_pixels = zeros(height, width);
    RGB_image = double(RGB_image);
    
    %count leaf pixels in image -> leaf_pixel_count
    %in leaf_pixels, assign a 1 to leaf pixels
    for row = 1:1:height
        for column = 1:1:width
            total = sum(RGB_image(row, column, :));
            if total ~= 765 && total ~= 0
                leaf_pixels(row, column) = 1;
                leaf_pixel_count = leaf_pixel_count + 1;
            end
        end
    end
    
    %create new image for heatmap
    heatmap = zeros(height, width, pages);
    total_NAI = 0;
    RGB_image = uint8(RGB_image); %image must be in uint8 to change color space
    
    if strcmp(color_space, 'sRGB') == 1
        target_image = double(RGB_image);
    elseif strcmp(color_space, 'HSV') == 1
        target_image = rgb2hsv(RGB_image);
    elseif strcmp(color_space, 'YIQ') == 1 
        target_image = rgb2ntsc(RGB_image);
    elseif strcmp(color_space, 'L*a*b*') == 1
        target_image = rgb2lab(RGB_image);
    elseif strcmp(color_space, 'YCbCr') == 1
        target_image = double(rgb2ycbcr(RGB_image));
    else
        disp('Color space not recognized');
        return;
    end
    
    %apply regression to each leaf pixel to predict NAI at each pixel
    %assign pixel color in heatmap image based on predicted NAI
    for row = 1:1:height
        for column = 1:1:width
            if leaf_pixels(row, column) == 1
                pixel = reshape(target_image(row, column, :), [1, 3]);
                NAI = ExampleRegressionName.predictFcn(pixel); %replace with name of desired regression
                total_NAI = total_NAI + NAI;
                if NAI < 20
                    heatmap(row, column, :) = [231, 213, 217]; %pale white
                elseif NAI < 40
                    heatmap(row, column, :) = [223, 170, 186];
                elseif NAI < 60
                    heatmap(row, column, :) = [215, 127, 155];
                elseif NAI < 80
                    heatmap(row, column, :) = [207, 85, 124];
                elseif NAI < 100
                    heatmap(row, column, :) = [199, 42, 93];
                else
                    heatmap(row, column, :) = [191, 0, 63]; %dark pink
                end
            end
        end
    end
    
    %export heatmap image as a .png file to folder with original images
    %to name heatmap images, add 'heatmap_' to beginning of original image name
    heatmap = uint8(heatmap);
    filename = strcat(myFolder, '/Heatmap_', baseFileName);
    imwrite(heatmap, filename);
    
    %calculate average NAI by dividing total_NAI by number of leaf pixels
    avg_NAI{image_count + 1, 1} = cellstr(baseFileName);
    avg_NAI{image_count + 1, 2} = total_NAI/leaf_pixel_count;
end

clearvars -except avg_NAI

File: makeHeatmap.R
# Given an image of a leaf, create false color heatmap image based on predicted NAI
# Inputs: leaf image, and trained regression

makeHeatmap <- function(heatmapPath, transformParams, modelAccuracies, allModels){
  
  # use most accurate model
  modelName <- modelAccuracies$modelName[1]
  colorSpace <- modelAccuracies$modelColorSpace[1]
  regressionModel <- allModels[[colorSpace]][[modelName]]
  
  # get list of image names in folder
  imageNames <- list.files(heatmapPath)
  
  # initialize vector to store average NAI values for each image
  averageNAI <- vector()
  
  # loop over all images in folder specified by imagePath
  for(i in 1:length(imageNames)){
    print(paste("Now making heatmap of image", imageNames[i]))
    leafImage <- readImage(paste(heatmapPath, "/", imageNames[i], sep=""))
    
    # initialize empty matrix to store color information for heatmap image
    heatmap <- array(0, dim=c(nrow(leafImage), ncol(leafImage), 4))
    
    # convert image into appropriate color space to match regression model, and center and scale pixel values
    if(colorSpace != "sRGB"){
      conLeafImage <- array(0, dim=c(nrow(leafImage), ncol(leafImage), 4))
      for(row in 1:nrow(leafImage)){
        for(col in 1:ncol(leafImage)){
          if(leafImage[row, col, 4] == 1){
            conLeafImage[row, col, 4] <- 1
            sRGBpixel <- data.frame(R=leafImage[row, col, 1], G=leafImage[row, col, 2], B=leafImage[row, col, 3])
            # color space conversion
            convertedPixel <- convertColorSpace(sRGBpixel, colorSpace)
            # centering and scaling
            convertedPixel <- predict(transformParams[[colorSpace]], convertedPixel)
            # insert converted pixel into image
            conLeafImage[row, col, 1:3] <- convertedPixel[1:3]
          }
        }
      }
    }
    # if model is in sRGB color space, no color space conversion necessary, but centering and scaling still needed
    else{
      conLeafImage <- array(0, dim=c(nrow(leafImage), ncol(leafImage), 4))
      for(row in 1:nrow(leafImage)){
        for(col in 1:ncol(leafImage)){
          if(leafImage[row, col, 4] == 1){
            conLeafImage[row, col, 4] <- 1
            sRGBpixel <- data.frame(R=leafImage[row, col, 1], G=leafImage[row, col, 2], B=leafImage[row, col, 3])
            # centering and scaling
            convertedPixel <- predict(transformParams[[colorSpace]], sRGBpixel)
            # insert converted pixel into image
            conLeafImage[row, col, 1] <- convertedPixel[[1]]
            conLeafImage[row, col, 2] <- convertedPixel[[2]]
            conLeafImage[row, col, 3] <- convertedPixel[[3]]
          }
        }
      }
    }
    
    allNAI <- vector()
    
    # make prediction of NAI with model based on color data, and assign appropriate color to heatmap matrix
    for(row in 1:nrow(conLeafImage)){
      for(col in 1:ncol(conLeafImage)){
        if(conLeafImage[row, col, 4] == 1){
          heatmap[row, col, 4] <- 1
          leafPixel <- conLeafImage[row, col, 1:3]
          if(colorSpace=="sRGB"){
            predNAI <- predict(regressionModel, cbind.data.frame(R=leafPixel[1], G=leafPixel[2], B=leafPixel[3]))
          }else if(colorSpace=="HSV"){
            predNAI <- predict(regressionModel, cbind.data.frame(H=leafPixel[1], S=leafPixel[2], V=leafPixel[3]))
          }else if(colorSpace=="YIQ"){
            predNAI <- predict(regressionModel, cbind.data.frame(Y=leafPixel[1], I=leafPixel[2], Q=leafPixel[3]))
          }else if(colorSpace=="YCbCr"){
            predNAI <- predict(regressionModel, cbind.data.frame(Y=leafPixel[1], Cb=leafPixel[2], Cr=leafPixel[3]))
          }else{
            predNAI <- predict(regressionMode, cbind.data.frame(L=leafPixel[1], a=leafPixel[2], b=leafPixel[3]))
          }
          
          # create vector of NAI values for each leaf pixel
          allNAI <- append(allNAI, predNAI)
          
          # assign false color to heatmap image pixel based on predicted NAI
          if(predNAI < 20){
            heatmap[row, col, 1:3] <- c(231, 213, 217)/255
          }else if(predNAI < 40){
            heatmap[row, col, 1:3] <- c(223, 170, 186)/255
          }else if(predNAI < 60){
            heatmap[row, col, 1:3] <- c(215, 127, 155)/255
          }else if(predNAI < 80){
            heatmap[row, col, 1:3] <- c(207, 85, 124)/255
          }else if(predNAI < 100){
            heatmap[row, col, 1:3] <- c(199, 42, 93)/255
          }else{
            heatmap[row, col, 1:3] <- c(191, 0, 63)/255
          }
        }
      }
    }
    writeImage(heatmap, paste(heatmapPath,  "/", strsplit(imageNames[i], split=".", fixed=TRUE)[[1]][[1]], "Heatmap.png", sep=""))
    averageNAI <- append(averageNAI, mean(allNAI))
  }
  return(data.frame(averageNAI=averageNAI, row.names=imageNames)) # need to return average predicted NAI for all images analyzed
}